portalnet/src/offer_queue.rs

```diff
@@ -0,0 +1,313 @@
+                    content_key,
+                    peer.node_id(),
+                );
+                return false;
+            }
+            if !node_ids.fallback.contains(peer) {
+                debug!(
+                    "Content key: {} already in offer queue, adding peer to fallback list: {}",
+                    content_key,
+                    peer.node_id()
+                );
+                node_ids.fallback.push(peer.clone());
+            }
+            self.content_key_map.insert(content_key.clone(), node_ids);
+            return false;
+        }
+        self.content_key_map.insert(
+            content_key.clone(),
+            SeenPeers {
+                origin: peer.clone(),
+                fallback: vec![],
+            },
+        );
+        true
+    }
+
+    pub fn successful_processing(&mut self, content_key: &TContentKey) {
+        debug!("Content key: {content_key} successfully processed");
+        self.content_key_map.remove(content_key);
+    }
+
+    pub fn failed_to_accept(&mut self, content_keys: &Vec<TContentKey>) {
+        for content_key in content_keys {
+            self.failed_to_process(content_key);
+        }
+    }
+
+    pub fn failed_to_process(&mut self, content_key: &TContentKey) {
+        if let Some(mut node_ids) = self.content_key_map.remove(content_key) {
+            if node_ids.fallback.is_empty() {
+                debug!(
+                    "Failed to process content key: {content_key}, but no fallback peers found."
+                );
+                return;
+            }
+            // choose a random seen peer as the fallback peer
+            // and send a FINDCONTENT request to them
+            node_ids.fallback.shuffle(&mut thread_rng());
+            let fallback_peer = node_ids.fallback.remove(0);
+            let request = Request::FindContent(FindContent {
+                content_key: content_key.to_bytes(),
+            });
+            debug!(
+                "Failed to process content key: {}, sending fallback FINDCONTENT to {}",
+                content_key,
+                fallback_peer.node_id()
+            );
+            let direction = RequestDirection::Outgoing {
+                destination: fallback_peer.clone(),
+            };
+            if let Err(err) = self
+                .command_tx
+                .send(OverlayCommand::Request(OverlayRequest::new(
+                    request, direction, None, None, None,
+                )))
+            {
+                warn!(
+                    "Failed to send FINDCONTENT request from offer queue to fallback peer: {}, error: {}",
+                    fallback_peer.node_id(),
+                    err
+                );
+            }
+        } else {
+            warn!(
+                "Failed to process content key: {content_key}, but a corresponding OfferQueue record was not found"
+            );
+        }
+    }
+}
```

Comment:
Shouldn't this have
```rust
        self.content_key_map.insert(
            content_key.clone(),
            SeenPeers {
                origin: fallback_peer.clone(),
                fallback: node_ids.fallback
            },
        );
```

Or is it intentional to only after attempt 1 more time after a failure occured

Reply:
Oh nvm I see how my statement is wrong. But shouldn't we try more then 1 fallback node if the first one fails

Reply:
My thinking is that it's just simpler to erase the record from the queue. The queue will submit a `FINDCONTENT` request to the fallback node. This request won't use the `OfferQueue` anyways, and I'm not sure it's worth the complexity to track a content key beyond submitting the `FINDCONTENT` request to the overlay.

I guess the one case that this doesn't handle, is that the `OfferQueue` will then mark incoming offers for the content key as ones we should accept, during the execution of the `FINDCONTENT` request, when technically we don't want to mark these as valid.

It wouldn't be too hard to just send a `callback` channel along with the `FINDCONTENT` request, in which case we could repeat this process, with all the fallback peers, until we've guaranteed a successful `FINDCONTENT` request.

But, I just don't think it's worth the complexity right now. At the very worst, we'll have 2 concurrent incoming processes for the same content key, but this is only for a somewhat unique edge case. It's still an improvement over the current process. And, if we continue to see a lot of those `Unable to store accepted content, content already stored` errors, then we can improve the `OfferQueue` to handle this / any other edge case.

Reply:
I am having trouble finding where we handle the accepted content from the FC you are sending in your pr
Yeah currently if the first one fails and a FC happens it isn't handled.

Currently this is sending a FC and not handling the response so even if the fallback has the data we won't get it

Reply:
iiuc, inside `OverlayService` we handle `FindContent` requests via `OverlayService::process_received_content()` which will validate & store the received content. Imo, we don't really care about the actual data at this level, as long as it was validated & stored locally. So there's no need to return it via a callback

Reply:
But that is only if you create the overlay request using `OverlayCommand::FindContentQuery` where as here you are using `OverlayCommand::Request`

Reply:
A FC Request command takes a channels for a responder and calls `OverlayService::process_content()` which only marks a query if provided as complete

This isn't a query and we don't pass a channel for a response hence we aren't handling the result

Reply:
I'm not positive... But if what you're saying is correct, then that means we aren't doing any validation on content received from a simple `FindContent` query initiated over jsonrpc? iiuc, that's essentially what I'm doing here by using the `OverlayCommand::Request` wrapper rather than the `OverlayCommand::FindContentQuery` wrapper. I'll dig more into this tomorrow and confirm/deny

Reply:
Ahhh, ok yeah, looks like I had some invalid assumptions going on here. 

Well, the problem with submitting a `OverlayCommand::FindContentQuery` is that it doesn't accept a target `enr`. eg. it just starts the recursive find content lookup process. I felt like it was better to submit a `FindContent` request to a peer who has offered us that content key rather than starting a RFC lookup. 

I'd also like to avoid using `OverlayCommand::Request` and processing the content received via a callback here inside `OfferQueue` (eg doing the validation / storing). But maybe that's not as thorny as it seems...

Reply:
Maybe the best case forword is to refactor `OverlayService::process_content()` and `Overlay::send_find_content()`

> I'd also like to avoid using OverlayCommand::Request and processing the content received via a callback here inside OfferQueue (eg doing the validation / storing). But maybe that's not as thorny as it seems...

I think we should use `OverlayCommand::Request` now that you mentioned

> Well, the problem with submitting a OverlayCommand::FindContentQuery is that it doesn't accept a target enr. eg. it just starts the recursive find content lookup process

We could use tokio::spawn to make it non blocking so `OverlayCommand::Request` with a reponse channel would work then

Reply:
Maybe one route would be if no responder is provider handle content in  `OverlayService::process_content()`

Reply:
I think providing `OverlayCommand::Request` a response channels calling `tokio::spawn` then handling might be the play though cause then we  could configure how many fallbacks we want to try.

I guess there are quite a few ways we can fix this issue